src/utils/number.ts

```diff
@@ -1,4 +1,97 @@
 import { numberKeys } from "./key";
+import { createLocaleNumberFormatter, getDecimalSeparator, getMinusSign } from "./locale";
+
+// regex for number sanitization
+const allLeadingZerosOptionallyNegative = /^([-0])0+(?=\d)/;
+const decimalAtEndOfStringButNotStart = /(?!^\.)\.$/;
+const allHyphensExceptTheStart = /(?!^-)-/g;
+const isNegativeDecimalOnlyZeros = /^-\b0\b\.?0*$/;
+
+// adopted BigInt solution from https://stackoverflow.com/a/66939244
+export class BigDecimal {
+  value: bigint;
+
+  // BigInt("0") === BigInt("-0") which strips the minus sign when typing numbers like -0.1
+  // https://forum.kirupa.com/t/js-tip-of-the-day-negative-zero/643094
+  isNegative: boolean;
+
+  // Configuration: constants
+  static DECIMALS = 20; // number of decimals on all instances
```

Comment:
How many decimals do we want to allow before rounding? It can be 50 or more if we really want

Reply:
Yeah, 100 sounds fine, might as well max it out as large as we think would accommodate all use cases.